mathjax3-ts/output/chtml/FontData.ts

```diff
@@ -0,0 +1,328 @@
+        big_op_spacing2:  .167,
+        big_op_spacing3:  .2,
+        big_op_spacing4:  .45, // .6,  // better spacing for under arrows and braces
+        big_op_spacing5:  .1,
+
+        surd_height:      .075,
+
+        scriptspace:         .05,
+        nulldelimiterspace:  .12,
+        delimiterfactor:     901,
+        delimitershortfall:   .3,
+
+        min_rule_thickness:  1.25     // in pixels
+    };
+
+    /*
+     * The default delimiter and character data
+     */
+    protected static defaultDelimiters: DelimiterMap = {};
+    protected static defaultChars: CharMapMap = {};
+
+    /*
+     * The default variants for the fixed size stretchy delimiters
+     */
+    protected static defaultSizeVariants: string[] = [];
+
+    /*
+     * The actual variant, delimiter, and size information for this font
+     */
+    protected variant: VariantMap = {};
+    protected delimiters: DelimiterMap = {};
+    protected sizeVariants: string[];
+
+    /*
+     * The actual font parameters for this font
+     */
+    public params: FontParameters;
+
+    /*
+     * Copies the data from the defaults to the instance
+     *
+     * @constructor
+     */
+    constructor() {
+        let CLASS = (this.constructor as typeof FontData);
+        this.params = Object.assign({}, CLASS.defaultParams);
+        this.sizeVariants = CLASS.defaultSizeVariants.slice(0);
+        this.createVariants(CLASS.defaultVariants);
+        this.defineDelimiters(CLASS.defaultDelimiters);
+        for (const name of Object.keys(CLASS.defaultChars)) {
+            this.defineChars(name, CLASS.defaultChars[name]);
+        }
+    }
+
+    /*
+     * Creates the data structure for a variant (an object with prototype chain
+     *   that includes a copy of the linked variant, and then the inherit variant chain.
+     *   The linked copy is updated automatically when the link variant is modified.
+     *   (The idea is to be able to have something like bold-italic inherit from both
+     *   bold and intalic by having the prototype chain include a copy of bold plus
+     *   the full italic chain.  So if something is not defined explicitly in bold-italic,
+     *   it defaults first to a bold version, than an italic version, then the normal
+     *   version, which is in the italic prototype chain.)
+     *
+     * @param{string} name     The new variant to create
+     * @param{string} inherit  The variant to use if a character is not in this one
+     * @param{string} link     A variant to search before the inherit one (but only
+     *                           its top-level object).
+     */
+    public createVariant(name: string, inherit: string = null, link: string = null) {
+        let variant = {
+            linked: [] as CharMap[],
+            chars: (inherit ? Object.create(this.variant[inherit].chars) : {}) as CharMap
+        };
+        if (link && this.variant[link]) {
+            Object.assign(variant.chars, this.variant[link].chars);
```

Comment:
It is a merge, and the properties are the character data arrays (the mapping from unicode number to CharData array).

The reason for this extra link is that for a `mathvariant` like `bold-italic`, you want to inherit from both the `bold` and `italic` variants, but the prototype chain can only inherit from one.  So for `bold-italic`, we make an object that has a prototype consisting of a *copy* of the `bold` data, and add the `italic` data as the prototype chain.  (Since this *is* a copy, we keep a record of this link so that if `bold` is changed later, we update this copy.  That is not needed for the prototype chain, since the prototypes are the actual objects, not copies.)  We then use this bold-plus-italic object as the prototype chain for the `bold-italic` object

That means that `bold-italic` will first look in its own object for specifically bold-italic glyphs that are defined there, then in the copy of the bold glyphs (only its top level is copied, not is prototype chain), and then the specifically italic glyphs, and then the prototype chain for italics, which is the normal glyphs.  Effectively, this means `bold-italic` looks for bold-italic, then bold, then italic, then normal glyphs in order to find the given character.

Something like `bold-script` will link the `script` font and use `bold` as its prototype.  So `bold-script` will look for specifically bold-script glyphs, then `script` glyphs, then `bold` glyphs, then `normal` ones.

This is set up to reduce the amount of duplication in the lists, as well as the amount of copying that must be done when new glyphs are added to variants that are in the default variants (e.g., normal, bold, italic).

Because `variant.chars` is an object that may have a prototype already, we can't use spread to merge the linked chars.  E.g., 

```
variant.chars = {...this.variant[link].chars};
```
would lose the prototype to the inherited characters.

Reply:
Thanks! This is a great explanation, and I believe should be kept for posterity.
Could you please make a comment out of \S 2+3?